### DISCUSSION TITLE
backend benchmark

### DISCUSSION THREAD

**Discussion by abdelaziz-mahdy** [serverpod/serverpod#1791] [Show and tell]
Hello Serverpod Community!

I've been working on a comprehensive benchmarking project to compare various backend technologies, including Serverpod, based on their speed and load capabilities. The goal is to assist developers in choosing the most suitable backend framework for their specific needs, focusing on real-world scenarios like write and read operations.

#### **Benchmark Overview:**
The benchmarks involve different backend technologies, including Python (Django), Dart (Serverpod), and JavaScript/TypeScript (Express with Node and Bun). I have used Locust for testing, simulating a scenario with 5000 users, a spawn rate of 10 users/second, and a duration of 500 seconds.

**Benchmarks Available Here:**  
[GitHub - Backend Benchmarks](https://github.com/zezo357/backend-benchmark)

**Why I Conducted These Benchmarks:**  
[GitHub Discussion on Serverpod](https://github.com/serverpod/serverpod/discussions/598)  
(Note: This discussion was started to address requests for Serverpod benchmarks.)

#### **Serverpod in Focus:**

- **Benchmark Graph for Serverpod:** 
![image](https://github.com/serverpod/serverpod/assets/25157308/fba7ee51-98dc-4691-8502-6cac180ed443)

- **Comparison with Other Technologies:**
![image](https://github.com/serverpod/serverpod/assets/25157308/19015603-f6e2-45b3-8aef-d00edd1c42a9)

#### **Discussion Points:**
- Serverpod's performance in write/read operations compared to other technologies.
- Insights into Serverpod's handling of high-load scenarios.
- Areas where Serverpod excels and where it has room for improvement.
- Why is Javascript faster.

#### **Call to Action:**
I'd love to invite the community to dive into these results, provide feedback, and discuss potential enhancements. Your insights and expertise can greatly contribute to improving Serverpod's performance and usability.

- Are there specific areas in Serverpod you think can be optimized?
- Any particular results that stand out to you?
- Suggestions and ideas for future benchmarks or tests.

This project is not just about comparing numbers; it's about understanding how we can collectively enhance the tools we rely on. I'm looking forward to a constructive and insightful discussion!

Thank you for your time and contributions!

**Comment by SandPod** [comment]
@zezo357 This is awesome! Thank you so much for putting this together! Looking forward to finding places where Serverpod can improve.

**Reply by abdelaziz-mahdy** [reply]
Hello,

here's what I have noticed:

1. **Database Read/Write Endpoints:**
   When calling endpoints that depend on database read and write operations, serverpod is slower than javascript and c#:
   ![Database Operation Issue](https://github.com/serverpod/serverpod/assets/25157308/efbdf3a3-b52b-46ec-a32e-28b7cf5aa61a)

2. **Static Endpoints:**
   Conversely, for static endpoints that merely return a string, the results are very close to each other:
   ![Static Endpoint Functionality](https://github.com/serverpod/serverpod/assets/25157308/144d8bce-9397-4528-b40d-122435a68a10)

These distinct outcomes in different endpoint types suggest a potential issue with the database interaction(due to lower cpu usage on database) or serialization process is taking longer than expected which may be the cause of the lower performance.

I am presenting this for your consideration and would appreciate any insights or feedback.

Thank you.
